### PR TITLE
[flink] fix false assumption on catalog table passing logic

### LINKED ISSUE #4372: [Bug] Assumption on the Consistency of Flink CatalogTable is Vulnerable
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.

### Paimon version

87321e13277bb32e909f4949bb0358eaffcf00e0

### Compute Engine

Flink

### Minimal reproduce step

Create a MaterializeTable and try to insert some data to it. When the code enters `FlinkTableFactory#createDynamicTableSink`, check `context.getCatalogTable().getOrigin()`.

### What doesn't meet your expectations?

The expected value in `context.getCatalogTable().getOrigin()` should be a `CatalogMaterializedTable`, while actually it is a `CatalogTable`.

### Anything else?

The cause of the unexpected behavior above is as follows: When creating a Flink job that inserts into Paimon(Reading from Paimon has the same problem), Flink will first invoke PaimonCatalog's `getTable` method to acquire certain metadata about the table, and then invoke `FlinkTableFactory#createDynamicTableSink` to create a flink sink. Suppose the table that Paimon returned in `getTable` is tableA and the table passed into `FlinkTableFactory#createDynamicTableSink` is tableB, Paimon has wrongly assumed that tableB == tableA or tableB == tableA.copy(...). However, Flink Catalog and TableFactory API has not guaranteed this assumption. During the optimization process of Flink SQL Planner, Flink might create another table and pass it to PaimonCatalog. Flink is not doing it wrong so long as the behaviors of the public APIs(CatalogTable) does not change.

This false assumption resulted in the following codes. You can see that for a table acquired through `context.getCatalogTable().getOrigin()`, Paimon assumes that the table must be a certain child class of CatalogBaseTable.

```java
// in AbstractFlinkTableFactory#createDynamicTableSource

if (origin instanceof SystemCatalogTable) {
    return new SystemTableSource(
            ((SystemCatalogTable) origin).table(),
            isStreamingMode,
            context.getObjectIdentifier());
} else {
```

```java
// in AbstractFlinkTableFactory#buildPaimonTable

if (origin instanceof DataCatalogTable) {
    FileStoreTable fileStoreTable = (FileStoreTable) ((DataCatalogTable) origin).table();
    table = fileStoreTable.copyWithoutTimeTravel(newOptions);
} else {
    table =
            FileStoreTableFactory.create(createCatalogContext(context))
                    .copyWithoutTimeTravel(newOptions);
}
```

Thus it is necessary to get these implementations rid of the false assumption to avoid potential correctness problems.

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by yunfengzhou-hub**
### Purpose

Linked issue: close #4372 

### Tests
 
The change is verified by FlinkCatalogTest.

### API and Format

This change does not affect API or Format.

### Documentation

This change does not introduce a new feature.

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java [hunk 1]
```diff
@@ -243,16 +257,34 @@ static Table buildPaimonTable(DynamicTableFactory.Context context) {
...
         newOptions.putAll(dynamicOptions);
 
-        // notice that the Paimon table schema must be the same with the Flink's
-        if (origin instanceof DataCatalogTable) {
-            FileStoreTable fileStoreTable = (FileStoreTable) ((DataCatalogTable) origin).table();
-            table = fileStoreTable.copyWithoutTimeTravel(newOptions);
+        FileStoreTable fileStoreTable;
+
+        // The following if conditions provide a shortcut to acquire Paimon table.
+        if (origin instanceof FormatCatalogTable) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:263]: It should not be a `FormatCatalogTable`.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java [hunk 2]
```diff
@@ -243,16 +257,34 @@ static Table buildPaimonTable(DynamicTableFactory.Context context) {
...
-                    FileStoreTableFactory.create(createCatalogContext(context))
-                            .copyWithoutTimeTravel(newOptions);
+            // In case the shortcut is not matched, the paimon table can still be acquired from
+            // catalog.
+            Identifier identifier =
+                    Identifier.create(
+                            context.getObjectIdentifier().getDatabaseName(),
+                            context.getObjectIdentifier().getObjectName());
+            try {
+                fileStoreTable = (FileStoreTable) flinkCatalog.catalog().getTable(identifier);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:279]: What do you want to solve? This should be a `CatalogMaterializedTable`?

Comment (yunfengzhou-hub) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:279]: The return type of `DynamicTableFactory.Context#getCatalogTable().getOrigin()` is `CatalogTable`, and `CatalogMaterializedTable` is not assignable from `CatalogTable`.

The problem I want to solve is that in the following code 
```java
if (origin instanceof DataCatalogTable) {
    FileStoreTable fileStoreTable = (FileStoreTable) ((DataCatalogTable) origin).table();
    table = fileStoreTable.copyWithoutTimeTravel(newOptions);
} else {
    table =
            FileStoreTableFactory.create(createCatalogContext(context))
                    .copyWithoutTimeTravel(newOptions);
}
```
Flink Paimon catalog assumed that if the origin table is not a `DataCatalogTable`, then it must support `FileStoreTableFactory#create` (i.e. the scheme of the origin table's path url can be matched with a FileIOLoader). While in practice I encountered a custom FileIO that does not support SPI dynamic loading, and thus would throw exceptions when running into `FileStoreTableFactory#create`.

**Comment by JingsongLi** [comment]
+1

**Pull request merged**